### PR TITLE
[TopOpt_in_PETSc] WIP: new builder

### PR DESCRIPTION
**Pull request by Wimmerer**
This shouldn't be merged in its current state. There is a pretty big problem:

I currently have to vendor PETSc. The reason is that the current PETSc builder clobbers lots of files when it builds for each set of scalars (real/complex, single/double, and Int32/Int64). In particular it clobbers all of `.h` files in `$includedir`, and all of the `$libdir/petsc` folder which contains configuration files.

The version required by this library `double_real_Int32` is built first, so only the `.so` file remains at the end. In order to have a dependency on PETSc these files need to be kept in some form. I'm just not sure how to do that correctly in the PETSc builder.

~~The second issue is that this repo has no license.~~ Found it hiding in plain sight.

### CODE REVIEW

**Comment by Wimmerer** [comment]
@vchuravy do you have any thoughts on this?

I can also try _not_ using PETSc's build system to build TopOpt, and just manually edit the compiler calls. That would be fragile (they recommend you _don't_ do that), but I could specify the location of the right files.

That still doesn't solve the issue with the PETSc builder deleting the `.h` files from previous (different number type) builds.

**Comment by vchuravy** [comment]
I haven't looked at the PETSc builder in detail, maybe @jkozdon has ideas?

**Comment by jkozdon** [comment]
I think we may need to save `${PETSC_DIR}/${PETSC_ARCH}/include/petscconf.h` (and maybe the other `.h` files in the directory).

I'm fine with the `.h` files being saved in some way. I think that you'd need to do that after this:
https://github.com/JuliaPackaging/Yggdrasil/blob/71094abadc5d9925e7633ed13321e20e31b95acb/P/PETSc/build_tarballs.jl#L77-L85
in the PETSc build script. Not sure if something like:
```sh
    # add suffix to library name
    if [[ "${target}" == *-mingw* ]]; then
        # changing the extension from so to dll.
        mv ${prefix}/lib/libpetsc.so.*.*.* "${libdir}/libpetsc_${1}_${2}_${3}.${dlext}"
    elif [[ "${target}" == *-apple* ]]; then
        mv ${prefix}/lib/libpetsc.*.*.*.${dlext} "${libdir}/libpetsc_${1}_${2}_${3}.${dlext}"
    else
        mv ${prefix}/lib/libpetsc.${dlext}.*.*.* "${libdir}/libpetsc_${1}_${2}_${3}.${dlext}"
    fi
    mkdir "${libdir}/include_${1}_${2}_${3}"
    mv ${prefix}/include/*.h "${libdir}/include_${1}_${2}_${3}"
```
would work?

**Comment by Wimmerer** [comment]
I'll make a PR to try that. I forget, is there a way to test more than one builder locally?

Relatedly, what should I do about the libraries? Just manually build against `libpetsc_${1}_${2}...`? Or use `-lpetsc` and make a symlink to the required library? I'll have to sidestep PETSc's build variables/system in order to do the former.

**Comment by mohamed82008** [comment]
Sorry to bother you all again with this, but is there any hope this PR can be done in the near future?

**Comment by Wimmerer** [comment]
https://github.com/JuliaPackaging/Yggdrasil/pull/3801 needs to be finished up. In particular I need to figure out how to do what Elliot recommended there towards the bottom. I have a few more questions about that, I'll ping him on Slack. Once that's done this should just work like normal. 

Sorry about the long gap there! The PETSc build system is a little frustrating to work around.

**Comment by mohamed82008** [comment]
No worries, I appreciate your help :)

**Comment by mohamed82008** [comment]
With #3801 merged, is this PR not blocked anymore?

**Comment by Wimmerer** [comment]
I have to finish it up but no it's not blocked anymore. I'll try to finish it tonight

**Comment by Wimmerer** [comment]
I expect almost all of those are red from using patchelf right now.

@giordano this will now "compile" but running `./topopt` within the builder results in:

```
Inconsistency detected by ld.so: dl-call-libc-early-init.c: 37: _dl_call_libc_early_init: Assertion `sym != N86_64-linux-gnu/sys-root/lib -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/lib -Wl,-rpath,/opt/x86_64-lin
``` 
(it gets cut off I suppose)

There's two questions here:
1. Will I be forced to use patchelf so long as the SONAME is still `libpetsc.so.3.16`?
2. What does this error even mean? It looks like perhaps topopt could not dynamically link to `libpetsc_double_real_Int32.so`? This is the function where the assertion fails:

```
   25 void
   26 _dl_call_libc_early_init (struct link_map *libc_map, _Bool initial)
   27 {
   28   /* There is nothing to do if we did not actually load libc.so.  */
   29   if (libc_map == NULL)
   30     return;
   31 
   32   const ElfW(Sym) *sym
   33     = _dl_lookup_direct (libc_map, "__libc_early_init",
   34                          0x069682ac, /* dl_new_hash output.  */
   35                          "GLIBC_PRIVATE",
   36                          0x0963cf85); /* _dl_elf_hash output.  */
   37   assert (sym != NULL);
   38   __typeof (__libc_early_init) *early_init
   39     = DL_SYMBOL_ADDRESS (libc_map, sym);
   40   early_init (initial);
   41 }
```
Is this even a valid test of whether `./topopt` will work on a users machine.

File: T/TopOpt_in_PETSc/bundled/Makefile [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+LDLIBS= -lpetsc_double_real_Int32 -lopenblas -lpthread -lm -lmpifort -lmpi -lstdc++ -ldl -lgfortran -lm -lgfortran -lm -lgcc_s -lstdc++ -ldl
+topopt: main.o TopOpt.o LinearElasticity.o MMA.o Filter.o PDEFilter.o MPIIO.o
+	$(CC) $(CPPFLAGS) -v -o topopt main.o TopOpt.o LinearElasticity.o MMA.o Filter.o PDEFilter.o MPIIO.o $(LDFLAGS) $(LDLIBS)
+
+libtopopt: TopOpt.o LinearElasticity.o MMA.o Filter.o PDEFilter.o MPIIO.o
+	$(CC) $(CPPFLAGS) -v -shared -o libtopopt.$(dlext) TopOpt.o LinearElasticity.o MMA.o Filter.o PDEFilter.o MPIIO.o $(LDFLAGS) $(LDLIBS)
+
+myclean:
+	rm -rf topopt *.o output* binary* log* makevtu.pyc Restart*
```
Comment (giordano) [inline T/TopOpt_in_PETSc/bundled/Makefile:14]: Missing line ending and trailing whitespace
```suggestion
	rm -rf topopt *.o output* binary* log* makevtu.pyc Restart*
```

File: T/TopOpt_in_PETSc/bundled/Makefile [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+CXXFLAGS = -fPIC
+CPPFLAGS=-I. -I$(includedir) -I$(PETSC_DIR)/include -I$(PETSC_DIR)/
+LDFLAGS= -Wl,-rpath,/workspace/destdir/lib/petsc/double_real_Int32/lib -L/workspace/destdir/lib/petsc/double_real_Int32/lib -Wl,-rpath,/workspace/destdir/lib -L/workspace/destdir/lib -Wl,-rpath,/opt/x86_64-linux-gnu/x86_64-linux-gnu/lib64 -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/lib64 -Wl,-rpath,/opt/x86_64-linux-gnu/lib/gcc/x86_64-linux-gnu/9.1.0 -L/opt/x86_64-linux-gnu/lib/gcc/x86_64-linux-gnu/9.1.0 -Wl,-rpath,/opt/x86_64-linux-gnu/lib/gcc -L/opt/x86_64-linux-gnu/lib/gcc -Wl,-rpath,/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/lib64 -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/lib64 -Wl,-rpath,/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/lib64 -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/lib64 -Wl,-rpath,/opt/x86_64-linux-gnu/x86_64-linux-gnu/lib -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/lib -Wl,-rpath,/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/lib -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/lib -Wl,-rpath,/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/lib -L/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/lib
```
Comment (giordano) [inline T/TopOpt_in_PETSc/bundled/Makefile:5]: There are _**far**_ too many rpaths set here.  Use _**only**_ what you need, not a single rpath extra.

File: T/TopOpt_in_PETSc/bundled/Makefile [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+PETSC_ARCH=x86_64-linux-gnu_double_real_Int32
+CXXFLAGS = -fPIC
+CPPFLAGS=-I. -I$(includedir) -I$(PETSC_DIR)/include -I$(PETSC_DIR)/
+LDFLAGS= -Wl,-rpath,/workspace/destdir/lib/petsc/double_real_Int32/lib -L/workspace/destdir/lib/petsc/double_real_Int32/lib
+LDLIBS= -lpetsc_double_real_Int32 -lopenblas -lpthread -lm -lmpifort -lmpi -lstdc++ -ldl -lgfortran -lm -lgfortran -lm -lgcc_s -lstdc++ -ldl
+topopt: main.o TopOpt.o LinearElasticity.o MMA.o Filter.o PDEFilter.o MPIIO.o
+	$(CC) $(CPPFLAGS) -v -o topopt main.o TopOpt.o LinearElasticity.o MMA.o Filter.o PDEFilter.o MPIIO.o $(LDFLAGS) $(LDLIBS)
+
+libtopopt: TopOpt.o LinearElasticity.o MMA.o Filter.o PDEFilter.o MPIIO.o
+	$(CC) $(CPPFLAGS) -v -shared -o libtopopt.$(dlext) TopOpt.o LinearElasticity.o MMA.o Filter.o PDEFilter.o MPIIO.o $(LDFLAGS) $(LDLIBS)
```
Comment (giordano) [inline T/TopOpt_in_PETSc/bundled/Makefile:11]:
```suggestion
	$(CC) $(CPPFLAGS) -v -o topopt $^ $(LDFLAGS) $(LDLIBS)

libtopopt.$(dlext): TopOpt.o LinearElasticity.o MMA.o Filter.o PDEFilter.o MPIIO.o
	$(CC) $(CPPFLAGS) -v -shared -o libtopopt.$(dlext) $^ $(LDFLAGS) $(LDLIBS)
```

File: T/TopOpt_in_PETSc/bundled/Makefile [hunk 4]
```diff
@@ -0,0 +1,14 @@
+PETSC_DIR=$(prefix)/lib/petsc/double_real_Int32
+PETSC_ARCH=x86_64-linux-gnu_double_real_Int32
+CXXFLAGS = -fPIC
+CPPFLAGS=-I. -I$(includedir) -I$(PETSC_DIR)/include -I$(PETSC_DIR)/
+LDFLAGS= -Wl,-rpath,/workspace/destdir/lib/petsc/double_real_Int32/lib -L/workspace/destdir/lib/petsc/double_real_Int32/lib
```
Comment (giordano) [inline T/TopOpt_in_PETSc/bundled/Makefile:5]:
```suggestion
LDFLAGS= -Wl,-rpath,${libdir}/petsc/double_real_Int32/lib -L${libdir}/petsc/double_real_Int32/lib
```

**Comment by giordano** [comment]
Ok, the problem is that `relpath(new_libdir, dirname(path))` at https://github.com/JuliaPackaging/BinaryBuilder.jl/blob/f20e6d0b9ddd66a097fa4a0beab70c2ea2f9b704/src/auditor/dynamic_linkage.jl#L440 returns
```
julia> relpath("/workspace/destdir/lib/petsc/double_real_Int32/lib", "/workspace/destdir/lib")
"petsc/double_real_Int32/lib"
```
and so [`normalize_rpath`](https://github.com/JuliaPackaging/BinaryBuilder.jl/blob/f20e6d0b9ddd66a097fa4a0beab70c2ea2f9b704/src/auditor/dynamic_linkage.jl#L405-L413) gets the path completely wrong.

File: T/TopOpt_in_PETSc/bundled/Makefile [hunk 5]
```diff
@@ -0,0 +1,14 @@
+PETSC_DIR=$(prefix)/lib/petsc/double_real_Int32
+PETSC_ARCH=x86_64-linux-gnu_double_real_Int32
+CXXFLAGS = -fPIC
+CPPFLAGS=-I. -I$(includedir) -I$(PETSC_DIR)/include -I$(PETSC_DIR)/
+LDFLAGS= -Wl,-rpath-link,${libdir}/petsc/double_real_Int32/lib -L${libdir}/petsc/double_real_Int32/lib
```
Comment (giordano) [inline T/TopOpt_in_PETSc/bundled/Makefile:5]:
```suggestion
LDFLAGS= -Wl,-rpath,${libdir}/petsc/double_real_Int32/lib -L${libdir}/petsc/double_real_Int32/lib
```

**Comment by giordano** [comment]
I can confirm this is fixed by https://github.com/JuliaPackaging/BinaryBuilder.jl/pull/1186.

File: T/TopOpt_in_PETSc/build_tarballs.jl [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+# Bash recipe for building across all platforms
+# New makefiles added, the patches fix some weird include issues mostly.
+# There is likely a better way to fix them, or upstream the fixes.
+script = raw"""
+
+cd TopOpt_in_PETSc
+cp ../Makefile Makefile
+make libtopopt
+make topopt
```
Comment (giordano) [inline T/TopOpt_in_PETSc/build_tarballs.jl:22]:
```suggestion
make libtopopt.${dlext}
make topopt${exeext}
```

File: T/TopOpt_in_PETSc/bundled/Makefile [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+PETSC_ARCH=x86_64-linux-gnu_double_real_Int32
+CXXFLAGS = -fPIC
+CPPFLAGS=-I. -I$(includedir) -I$(PETSC_DIR)/include -I$(PETSC_DIR)/
+LDFLAGS= -Wl,-rpath,${libdir}/petsc/double_real_Int32/lib -L${libdir}/petsc/double_real_Int32/lib
+LDLIBS= -lpetsc_double_real_Int32 -lopenblas -lpthread -lm -lmpifort -lmpi -lstdc++ -ldl -lgfortran -lm -lgfortran -lm -lgcc_s -lstdc++ -ldl
+topopt: main.o TopOpt.o LinearElasticity.o MMA.o Filter.o PDEFilter.o MPIIO.o
+	$(CC) $(CPPFLAGS) -v -o topopt $^ $(LDFLAGS) $(LDLIBS)
+
+libtopopt.$(dlext): TopOpt.o LinearElasticity.o MMA.o Filter.o PDEFilter.o MPIIO.o
+	$(CC) $(CPPFLAGS) -v -shared -o libtopopt.$(dlext) $^ $(LDFLAGS) $(LDLIBS)
```
Comment (giordano) [inline T/TopOpt_in_PETSc/bundled/Makefile:11]:
```suggestion
topopt$(exeext): main.o TopOpt.o LinearElasticity.o MMA.o Filter.o PDEFilter.o MPIIO.o
	$(CC) $(CPPFLAGS) -v -o $@ $^ $(LDFLAGS) $(LDLIBS)

libtopopt.$(dlext): TopOpt.o LinearElasticity.o MMA.o Filter.o PDEFilter.o MPIIO.o
	$(CC) $(CPPFLAGS) -v -shared -o $@ $^ $(LDFLAGS) $(LDLIBS)
```

**Comment by Wimmerer** [comment]
Thanks a bunch @giordano, so fast I didn't have a chance to fix! How long will it take to merge the fix to BB?

File: T/TopOpt_in_PETSc/bundled/Makefile [hunk 8]
```diff
@@ -0,0 +1,14 @@
+PETSC_DIR=$(prefix)/lib/petsc/double_real_Int32
```
Comment (giordano) [inline T/TopOpt_in_PETSc/bundled/Makefile:1]:
```suggestion
PETSC_DIR=$(libdir)/petsc/double_real_Int32
```

File: T/TopOpt_in_PETSc/bundled/Makefile [hunk 9]
```diff
@@ -0,0 +1,20 @@
+PETSC_DIR = $(libdir)/petsc/double_real_Int32
+PETSC_ARCH = x86_64-linux-gnu_double_real_Int32
```
Comment (giordano) [inline T/TopOpt_in_PETSc/bundled/Makefile:2]: What's this used for?

**Comment by giordano** [comment]
All green now :slightly_smiling_face:

File: T/TopOpt_in_PETSc/bundled/Makefile (hunk 9 above)
Comment (Wimmerer) [reply T/TopOpt_in_PETSc/bundled/Makefile:2]: It's a relic from using the PETSc build system. I'll remove it shortly, then we can merge! Thanks :)

**Pull request merged**